Working w/Python var in dev and prod (#17)
viam-labs/viamcodelabs: site/viamguides/src/environment-variables/environment-variables.md

    ```bash
    mkdir viam-rover
    cd viam-rover
    ```

### Prepare a virtual environment
[Set up a Python virtual environment](https://docs.viam.com/sdks/python/python-venv/) to avoid any conflicts with other projects or your system.
1. From the command line in your terminal window, install [`virtualenv` using `pip`](https://virtualenv.pypa.io/en/latest/installation.html#via-pip)
    ```bash
    python3 -m pip install --user virtualenv
    ```
1. **Create and activate a virtual environment**: In the project directory `viam-rover`, create and activate a virtual environment called `.venv` for Python to run in.
    ```bash
    python3 -m venv .venv
    source .venv/bin/activate
    ```
Now, `.venv` prepends the commands in your terminal window to indicate the Python packages being used are from this particular environment. You can exit this environment by running `deactivate`.

### Install the Python SDK

Install the Viam Python SDK and all required general dependencies.

1. **Install the SDK**: From the command line in your terminal window (inside the activated `.venv` Python environment), install the Viam Python SDK:
    ```bash

>     pip3 install viam-sdk

Reviewer: Once inside an activated virtual environment, it is typical to use the standard `pip` or `python` commands rather than the `pip3` or `python3` versions. 

```suggestion
    pip install viam-sdk
```

    > Windows is not supported. If you are using Windows, use the [Windows Subsystem for Linux (WSL)](https://learn.microsoft.com/en-us/windows/wsl/install) and install the Python SDK using the preceding instructions for Linux. For other unsupported systems, see [Installing from source](https://python.viam.dev/#installing-from-source).

<!-- ------------------------ -->
## Using variables in local development
Duration: 5

### Copy the sample code
1. **Refer to Python sample code**: In the Viam app, go to the **CONNECT** tab, and select **Python**. 
  ![Connect tab](assets/Connect.png)
1. **Show secrets**: The sample code shows you how to authenticate and connect to a machine, as well as some of the methods you can use on your configured components and services. To show your machine’s API key and API key ID in the sample code, toggle **Include secret** on the CONNECT tab’s Code sample page. 
  ![Code sample](assets/Code.png)
1. **Hide secrets**: Instead of leaving our sensitive credentials hard coded in this code sample, let's set up our API key and API key ID as environment variables. Toggle **Include secrets** once again, this time to remove the credentials from the code sample. 

> 1. **Copy the code**: From the command line in your terminal window, create a new file called `square.py`, and copy the code sample into this file.

Reviewer: This almost makes it sound like `touch square.py` is going to copy the contents of the code sample into the file. 

Maybe one step should be making the `square.py` file (why is it called that btw?), followed by a step to instruct copying the contents of the code sample to that new file.

Author: Separated into 2 steps, adding in a different commit.

I was following a different tutorial, but then simplified to remove the "drive rover in a square" code. I updated to `rover.py` file name, in a different commit.

   ```bash
   touch square.py
   ```
### Run the sample code
1. **Create `.env` file**: From the command line, create a new file called `.env` to store our environment variables to use in local development.
    ```bash
    touch .env
    ```
1. **Input environment variables**: Add your environment variables to the `.env` file, formatted like the following:
    ```
    ROBOT_API_KEY="sijdb24bjnxsmp18mij0hace6smihu0r"

>     ROBOT_API_KEY_ID="9e135013-f3bf-40fc-a0e3-e6fc66a418d7"

Reviewer: Should we call out to fill in the values for API-KEY and API-KEY-ID from the code sample as these environment variables instead of copying this explicitly?

    ```
1. **Install `python-dotenv` package**: From the command line, still within our virtual environment `.venv`, install the [`python-dotenv`](https://pypi.org/project/python-dotenv/) package to read key-value pairs from an `.env `file and can set them as environment variables.
    ```bash

>     pip3 install python-dotenv

Reviewer: ```suggestion
    pip install python-dotenv
```

   ```
1. **Load the variables into your environment**: At the top of your `square.py` file, add the following code to load variables into your environment.
    ```python
    from dotenv import load_dotenv
    import os

    load_dotenv()  # Loads the environment variables from the .env file

>     robot_api_key = os.getenv('ROBOT_API_KEY')
>     robot_api_key_id = os.getenv('ROBOT_API_KEY_ID')

Reviewer: It might be personal preference, but many of the relevant Python repos for Viam projects will "CONSTANT_CASE" these variable names to indicate they should remain constant during the lifetime of the program. It should also help reflect that these values came from an environment variable of the same name:

```suggestion
    ROBOT_API_KEY = os.getenv('ROBOT_API_KEY')
    ROBOT_API_KEY_ID = os.getenv('ROBOT_API_KEY_ID')
```

    ```
1. **Use variables in your code**: You can now use these variables within your code. On rows 7 and 8 in `square.py`, you load variables into the environment. And then on rows 20 and 21, you use those variables to access your machine.
  ![Code in VS Code](assets/vscode.png)

> 1. **Run the code:** Run the sample code to connect to your machine

Reviewer: What should the reader expect to see after running the code?

   ```bash

>    python3 square.py

Reviewer: Same feedback about using the un-suffixed command for Python when in a venv:

```suggestion
   python square.py
```

   ```

> If you want to play with the rover, go to the Viam app **CONTROL** tab to interact with your machine’s resources. Click on one of the camera panels and toggle the camera stream on so you can observe the rover’s movements. Or try adding more code to [drive the rover in a square](https://docs.viam.com/get-started/drive-rover/).

Reviewer: Should this be considered an "aside" callout?

In the next few sections, learn how to use variables on your machine in production.

<!-- ------------------------ -->
## Add the control code to your machine
Duration: 3

Once you have the code working in your local development environment, you may choose to move your control code to your machine, so that it can run in production. 

> aside positive
> If you followed [Try Viam](https://app.viam.com/try), you won't have access to add control code to the machine, since you are just borrowing the rover. However, you can skip to the next part to see how you can set up variables on the machine. If you're using your own wheeled robot, you can complete these steps.

### Add the control code to your machine

> 1. **SSH into your board**: From the terminal window, run the following command to [SSH (Secure Shell) into your board](https://docs.viam.com/installation/prepare/rpi-setup/#connect-with-ssh), where the text in `<>` should be replaced (including the `<` and `>` symbols themselves) with the `user` and `hostname` you configured when you set up your machine. 

Reviewer: On some networks, mDNS might not work as expected so the `hostname.local` alias would fail to resolve. Should we described a fallback to the static IP address shown in the Viam app just in case?

    ```bash
    ssh <USERNAME>@<REMOTE-HOSTNAME>.local
    ```
1. **Create a project directory**: Create a new directory on your machine to hold your project code.
    ```bash
    mkdir robot
    ```

> 1. **Install dependencies**: Install the Viam Python SDK (and other dependencies if required) into the folder. 

Reviewer: Newer versions of Debian (Bookworm and beyond), along with its derivatives (Ubuntu, Raspbian, etc) will require a virtual environment before installing any Python packages in order to prevent impacting any other global programs.

    ```bash
    pip3 install viam-sdk
    ```
<!-- TODO do i need to activate a venv? -->
<!-- TODO do i need to chmod? -->
1. **Copy file(s) to the machine**: From the original computer you were developing on (not the secure shell prompt accessing your machine), copy the `square.py` file to your machine.
    ```bash

>     scp square.py <USERNAME>@<REMOTE-HOSTNAME>.local:/home/myboard/robot/square.py

Reviewer: The default directory for scp is the same as when ssh-ing into a machine, which is the home directory of the username logging in. This means you can use a relative path after the `:` to make the command more copy-pastable:

```suggestion
    scp square.py <USERNAME>@<REMOTE-HOSTNAME>.local:./robot/square.py
```

    ```
1. **Add a process**: [Configure a process](https://docs.viam.com/configure/processes/#configure-a-process) by going to the **CONFIGURE** tab of the Viam app. Click the plus icon (**+**) next to your machine in the left-hand menu and select **Process**. Name your process `process-square`.
  ![Add a process](assets/process.png)

> 1. **Find executable path**: To find out the path of the Python 3 executable that is being used on your machine, you can input `which python3` in the command line from within your secure shell. You will need this output for the upcoming steps.

Reviewer: `which python3` should be run after activating the virtual environment under which the viam-sdk and other dependencies were installed.

## Using variables on machine in production
Duration: 2

Now that you've moved your control code to your machine, let's set up our environment variables to use in production.

### Update the configuration JSON

1. **Locate the modular resource**: In the Viam app under the **CONFIGURE** tab, select the modular resource in the side bar to locate the card where we will add environment variables. At the top right of the card, select the curly braces icon to see advanced configuration details.
  ![Config icon](assets/processCurly.png)
1. Add a new JSON property called `env` and your environment variables, formatted like the following.
  ![Add environment variables](assets/envVariables.png)
1. Save the changes at the top of the page, and restart the machine in the status dropdown at the top.
  ![restart the machine](assets/restart.png)

### Other ways to use variables 

This is the recommended approach to [using Viam environment variables within a a registry module](https://docs.viam.com/registry/configure/#use-environment-variables-with-a-registry-module), or other modular resources on a physical device. 

But there are other ways to use environment variables:

> - **Add virtual environment**: If you're working with modules, you can add a Python virtual environment to a [persistent folder location](https://docs.viam.com/registry/configure/#default-environment-variables) called `VIAM_MODULE_DATA`, where a module can store data across reboots and versions, after the module is first installed.

Reviewer: ```suggestion
- **Add virtual environment**: If you're working with modules, you can add a Python virtual environment to a [persistent folder location](https://docs.viam.com/registry/configure/#default-environment-variables) set to the environment variable `VIAM_MODULE_DATA`, where a module can store data across reboots and versions, after the module is first installed.
```

- **Update root profile**: If you have administrative (root) privileges, you can add environment variables to your machine's root profile (not the user profile). You'll need to modify the global configuration files that affect all users on the system.
  - Linux and MacOS users can update a root profile like at `/etc/profile`, `/etc/bash.bashrc`, or `/etc/zsh/zshrc`
  - Windows users can update **System Properties** → **Environment Variables** to set system-wide variables.

<!-- ------------------------ -->
## Next Steps
Duration: 2

> Knowing how to manage variables allows you to work more efficiently and keep your code DRY (Don't Repeat Yourself) for more maintainable, readable, and error free code. It's also important for security reasons not to leak sensitive data and risk access to secure systems.

Reviewer: ```suggestion
Knowing how to manage environment variables allows you to work more efficiently and keep your code DRY (Don't Repeat Yourself) for more maintainable, readable, and error free code. It's also important for security reasons not to leak sensitive data and risk access to secure systems.
```
